### PR TITLE
Merge of PRs #39 #41 and #42, update versions of stm32f7xx-hal and stm32f4xx-hal, convert CI to github actions

### PR DESCRIPTION
**Pull request by datdenkikniet**
This PR merges all changes and resolves the conflicts that occur for #39, #41, and #42, as all three of these changes are rather important and it would be really nice to get all of them in.

AFAICT I have merged the changes correctly, as everything compiles and the seemingly most important changes have been carried over.

Also updates to the latest available versions of `stm32f7xx-hal` and `stm32f4xx-hal`

Converts travis CI to github CI actions, courtesy of @korken89

### CODE REVIEW

**Comment by thalesfragoso** [comment]
Thanks for your work, I get it that seeing a lot of unmerged PRs isn't nice, but I fear that this type of PR ends up making the review process more difficult.

If you could separate the change to github actions to another PR that would be great, it's probably the most important change right now.

**Comment by korken89** [comment]
Hi @thalesfragoso !

Thanks for your comment. 
I would not see it like that, it was not trivial in which order these PRs needed to be merged and updated to the correct HALs to make it all work (I already saw that you had comments on this in the original F1 PR, which we have fixed).
Plus we have tested this combo PR on F7 hardware so far, so it's tested and working with the latest HAL. :tada: 
Moreover, the PRs are quite old and stale and we are on the ball and available to do fixes quickly to help speed this along.

However to give to an overview so you can look at each of the old PRs and still give comments here.
We have made it so that we merged the PRs as-is into this branch and then did the fixes.
This also means you can jump into the commits and it will 100% match the original PRs.

On the addition of GHA, I can make it its own PR, but if we merge this one there is no need.
But if you insist I can do it.

Hit me up on matrix if you want to discuss real-time!

File: examples/ip-f107.rs
Comment (thalesfragoso) [inline examples/ip-f107.rs:86]: Could you explain why you think the MCO is necessary ? Ping @franeklubi

File: src/lib.rs
Comment (thalesfragoso) [inline src/lib.rs:88]: This is not true for the F1 code.

File: src/setup.rs
Comment (thalesfragoso) [inline src/setup.rs:296]: As I mentioned in the other PR:

It seems that we're relying on the HAL to always set the pin at max speed 50MHz, is that right ? It seems that the HAL indeed does that, but is that a documented behavior ?

@korken89 Table 209 does not guarantee that the pins will be in the right speed, it requires it.

**Review by thalesfragoso — COMMENTED** [review]
My thanks to you and all the people involved, nice work.

I left a few questions and comments.

File: .github/workflows/build.yml
Comment (korken89) [reply .github/workflows/build.yml:43]: I kept this as the original CI built for x86. I did try with thumb as well, and both works.

File: .github/workflows/build.yml
Comment (korken89) [reply .github/workflows/build.yml:154]: Correct!

File: Cargo.toml
Comment (korken89) [reply Cargo.toml:23]: Yeah there are a lot of changes in the f7 HAL, so I already reached out here to try and get a release: https://github.com/stm32-rs/stm32f7xx-hal/issues/185
Maybe someone feels like helping out?

File: README.md
Comment (korken89) [reply README.md:75]: We could, the original API did not expose the MMC though. 
Do we want to add it in this PR or have that addition be its own PR?

File: examples/ip-f107.rs
Comment (korken89) [reply examples/ip-f107.rs:86]: This is not needed, but was part of the example. We can remove it IMO.

File: src/lib.rs
Comment (korken89) [reply src/lib.rs:88]: Which part?

File: src/setup.rs
Comment (korken89) [reply src/setup.rs:296]: Correct, it is made by the mode setting when making a pin `Output`, see here: https://github.com/stm32-rs/stm32f1xx-hal/blob/master/src/gpio.rs#L967

- `0b11` is output @ 50 Mhz
- `0b01` is output @ `0 Mhz
- `0b10` is output @ 2 Mhz

File: README.md
Comment (korken89) [reply README.md:75]: On this, if we add MMC we should probably add PTP as well.

Comment (thalesfragoso) [reply README.md:75]: What I meant is that this PR changed the `new` method to require the MMC to be passed in, so this example on the README wouldn't compile.

File: examples/ip-f107.rs
Comment (thalesfragoso) [reply examples/ip-f107.rs:86]: If it isn't needed then we should remove it, otherwise the users will think it's necessary.

File: src/lib.rs
Comment (thalesfragoso) [reply src/lib.rs:88]: "Automatically sets slew rate to very high"

File: src/setup.rs
Comment (thalesfragoso) [reply src/setup.rs:296]: Yes, but we should ask the maintainers of the F1 HAL to document this default behavior so we can rely on it, but we don't have to wait for it to merge this PR.

Either way, I think we should document this special requirement for the F1 in the new function.

File: Cargo.toml
Comment (datdenkikniet) [reply Cargo.toml]: @korken89 convinced the F7 people to create a new release, and version has been pinned to that new version now.

File: README.md
Comment (datdenkikniet) [reply README.md:69]: This example has been fixed, and we added a test that explicitly attempts to compile the examples given in the README as well

File: src/lib.rs
Comment (datdenkikniet) [reply src/lib.rs:88]: The slew rate is now automatically set to `VeryHigh` (or `MHz50`) for `stm32f107` as well

File: src/setup.rs
Comment (datdenkikniet) [reply src/setup.rs]: The slew rate is now automatically set to VeryHigh (or MHz50) for `stm32f107` as well

**Comment by korken89** [comment]
@thalesfragoso All your requested  changes are now ready

File: src/setup.rs
Comment (thalesfragoso) [inline src/setup.rs:305]: This is clever indeed. However, it's still unsound and a bit hard to make it sound.

We could use `cortex_m::interrupt::free` to make your assumption true, but that still leaves us with the `transmute`, which is sound right now (with `interrupt::free`), but the HAL can make a simple semver compatible change and make this unsound and we would be the ones to blame.

I see two alternatives:
1. Just go back to the documentation stating that the pin should be set to high speed, which is quite fine, given that it's the default behavior for the HAL today.
2. Use `interrupt::free` and change the registers manually ourselves (with the pac). This is quite cumbersome due to the several different pins involved.

I would be fine with either one.

Comment (datdenkikniet) [reply src/setup.rs]: Thanks for the look-over :)

I've added a critical section around the use of the `unsafe`ly construced `CRL` register now. I've also added a comment explaining why it should be impossible for a change to the f1 HAL to introduce unsound behaviour.

Comment (thalesfragoso) [reply src/setup.rs]: Right, this is good enough, we already use critical sections in other configurations, so we're already risking some problems if this initial configuration is done in an interrupt, which would be quite weird anyways.

**Review by thalesfragoso — APPROVED** [review]
Great work folks, sorry for taking this long.

**Pull request merged**
